drivers/network/netio/netio.c

```diff
@@ -0,0 +1,1515 @@
+{
+    PWSK_SOCKET_INTERNAL Socket = CONTAINING_RECORD(SocketParam, WSK_SOCKET_INTERNAL, Socket);
+    NTSTATUS status = STATUS_NOT_IMPLEMENTED;
+
+    FUNCTION_TRACE;
+
+    if (Socket == NULL)
+    {
+        status = STATUS_INVALID_PARAMETER;
+        goto err_out;
+    }
+    IoSetNextIrpStackLocation(Irp);
+
+    switch (RequestType)
+    {
+        case WskSetOption:
+            switch (Level)
+            {
+                case SOL_SOCKET:
+                    switch (ControlCode)
+                    {
+                        case SO_REUSEADDR:     /* add more supported flags here */
+                            if (InputBuffer == NULL)
+                            {
+                                DPRINT1("WskControlSocket: Need an InputBuffer for this operation\n");
+                                status = STATUS_INVALID_PARAMETER;
+                                break;
+                            }
+                            if (InputSize < sizeof(UINT))
+                            {
+                                DPRINT1("WskControlSocket: InputBuffer too small for this operation\n");
+                                status = STATUS_INVALID_PARAMETER;
+                                break;
+                            }
+                            UINT flag = *(PUINT)InputBuffer;
+                            if (flag != 0)
+                            {
+                                Socket->Flags |= ControlCode;
+                            }
+                            else
+                            {
+                                Socket->Flags &= ~ControlCode;
+                            }
+                            status = STATUS_SUCCESS;
+                            break;
+                            /* Windows specific. This sets the mask for callback functions: */
+                        case SO_WSK_EVENT_CALLBACK:
+                            if (InputBuffer == NULL)
+                            {
+                                DPRINT1("WskControlSocket: Need an InputBuffer for this operation\n");
+                                status = STATUS_INVALID_PARAMETER;
+                                break;
+                            }
+                            if (InputSize < sizeof(WSK_EVENT_CALLBACK_CONTROL))
+                            {
+                                DPRINT1("WskControlSocket: InputBuffer too small for this operation\n");
+                                status = STATUS_INVALID_PARAMETER;
+                                break;
+                            }
+                            PWSK_EVENT_CALLBACK_CONTROL Control = (WSK_EVENT_CALLBACK_CONTROL *)InputBuffer;
+
+                            if (((Socket->CallbackMask & WSK_EVENT_ACCEPT) == 0) &&
+                                ((Control->EventMask & WSK_EVENT_ACCEPT) == WSK_EVENT_ACCEPT))
+                            {
+                                Socket->CallbackMask = Control->EventMask;
```

Comment:
It looks like, because this is done basically unconditionally (in both blocks of the if), this could be moved outside and before the if.